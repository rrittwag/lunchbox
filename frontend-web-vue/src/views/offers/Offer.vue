<script setup lang="ts">
import type { LunchOffer } from '@/model/lunch'
import { formatEuro } from '@/util/formatting'
import Badge from '@/views/offers/Badge.vue'
import { computed } from 'vue'

const props = defineProps<{
  offer: LunchOffer
  showDetailsInXS?: boolean
}>()

const sortedTags = computed(() => props.offer.tags.slice().sort())
const priceAsString = computed(() => formatEuro(props.offer.price))
const isVeggie = (tagLabel: string) => tagLabel === 'vegetarisch' || tagLabel === 'vegan'
</script>

<template>
  <li class="w-full pt-3 sm:pt-4">
    <div class="flex w-full items-baseline">
      <h4 class="grow text-xl leading-tight text-neutral-800">
        {{ props.offer.name }}
      </h4>
      <span v-if="props.offer.price" aria-label="Preis" class="shrink-0 pl-2 text-xl leading-tight text-neutral-800">
        <small class="pl-2 pr-1 text-lg font-light text-neutral-800">â‚¬</small>{{ priceAsString }}
      </span>
    </div>

    <p
      role="note"
      class="pl-px pt-px font-light leading-snug text-neutral-600 sm:block"
      :class="props.showDetailsInXS ? 'flex' : 'hidden'"
    >
      {{ props.offer.description }}
    </p>

    <p class="pt-1 sm:flex" :class="props.showDetailsInXS ? 'flex' : 'hidden'">
      <Badge
        v-for="tagLabel in sortedTags"
        :key="tagLabel"
        role="note"
        :label="tagLabel"
        :color="isVeggie(tagLabel) ? 'bg-success-200' : 'bg-accent-200'"
      />
    </p>
  </li>
</template>
