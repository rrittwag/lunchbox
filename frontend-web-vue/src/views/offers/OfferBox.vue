<template>
  <article
    class="flex flex-col items-center rounded-lg border-l-8 border-primary-500 bg-card px-4 pt-2 pb-3 shadow-xl sm:border-l-0 sm:border-t-8 sm:pt-4 sm:pb-6"
    @click="onClick"
  >
    <h3 class="font-display text-3xl text-neutral-700 sm:text-4xl">
      {{ props.provider.name }}
    </h3>
    <ul class="w-full list-none">
      <Offer v-for="offer in props.offers" :key="offer.id" :offer="offer" :showDetailsInXS="showDetails" />
    </ul>
  </article>
</template>

<script setup lang="ts">
import { ref } from 'vue'
import Offer from '@/views/offers/Offer.vue'
import { LunchOffer, LunchProvider } from '@/model/lunch'

const props = defineProps<{
  provider: LunchProvider
  offers: LunchOffer[]
}>()

const showDetails = ref(false)
const onClick = () => (showDetails.value = !showDetails.value)
</script>
