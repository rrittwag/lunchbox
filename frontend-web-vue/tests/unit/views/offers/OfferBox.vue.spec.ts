import OfferBox from '@/views/offers/OfferBox.vue'
import { mensa, gyros, soljanka } from '@tests/unit/test-data'
import Offer from '@/views/offers/Offer.vue'
import { shallowMount } from '@vue/test-utils'

describe('OfferBox', () => {
  it('renders snapshot', () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [gyros, soljanka] },
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders title & offers', () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [gyros, soljanka] },
    })

    expect(wrapper.text()).toEqual(mensa.name)
    const items = wrapper.findAllComponents(Offer)
    expect(items.length).toEqual(2)
  })

  it('renders offers in order', () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [gyros, soljanka] },
    })

    const items = wrapper.findAllComponents(Offer)
    expect(items[0].props('offer')).toEqual(gyros)
    expect(items[1].props('offer')).toEqual(soljanka)
  })

  it('WHEN offers are empty  THEN render just title', () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [] },
    })

    expect(wrapper.text()).toEqual(mensa.name)
    const items = wrapper.findAllComponents(Offer)
    expect(items.length).toBe(0)
  })

  it('hide details for screen size XS', () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [gyros, soljanka] },
    })

    const items = wrapper.findAllComponents(Offer)
    expect(items[0].props('showDetailsInXS')).toEqual(false)
    expect(items[1].props('showDetailsInXS')).toEqual(false)
  })

  it('WHEN clicked  THEN show details for screen size XS', async () => {
    const wrapper = shallowMount(OfferBox, {
      props: { provider: mensa, offers: [gyros, soljanka] },
    })

    await wrapper.trigger('click')
    await wrapper.vm.$nextTick() // Wait until trigger events have been handled

    const items = wrapper.findAllComponents(Offer)
    expect(items[0].props('showDetailsInXS')).toEqual(true)
    expect(items[1].props('showDetailsInXS')).toEqual(true)
  })
})
