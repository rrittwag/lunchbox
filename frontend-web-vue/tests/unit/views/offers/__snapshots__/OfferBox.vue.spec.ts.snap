// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferBox renders snapshot 1`] = `
<article
  class="flex flex-col items-center px-4 pt-2 pb-3 sm:pt-4 sm:pb-6 shadow-xl bg-card rounded-lg border-l-8 sm:border-l-0 sm:border-t-8 border-primary-500"
>
  <h3
    class="text-3xl sm:text-4xl font-display text-neutral-700"
  >
    Mensa
  </h3>
  <ul
    class="w-full list-none"
  >
    
    <offer-stub
      offer="[object Object]"
      showdetailsinxs="false"
    />
    <offer-stub
      offer="[object Object]"
      showdetailsinxs="false"
    />
    
  </ul>
</article>
`;
